src/browser_action/vitals.js

```diff
@@ -45,6 +45,20 @@
   // Registry for badge metrics
   const badgeMetrics = initializeMetrics();
 
+  // Set up extension message port with the service worker
+  let port = chrome.runtime.connect();
+  port.onMessage.addListener((response) => {
+    drawOverlay(badgeMetrics, response.tabId);
```

Comment:
nit: it might make sense to be a bit more defensive here since these are slightly more decoupled now. Just something like

```suggestion
    if (response.tabId === undefined) {
      return;
    }

    drawOverlay(badgeMetrics, response.tabId);
```